src/Xamarin.Android.Tools.AndroidSdk/Sdks/AndroidSdkBase.cs

```diff
@@ -9,6 +9,14 @@ namespace Xamarin.Android.Tools
 {
 	abstract class AndroidSdkBase
 	{
+		const int MinimumCompatibleNDKMajorVersion = 16;
+		const int MaximumCompatibleNDKMajorVersion = 21;
+
+		static readonly char[] SourcePropertiesKeyValueSplit = new char[] { '=' };
+
+		// Per https://developer.android.com/studio/command-line/variables#envar
+		protected static readonly string[] AndroidSdkEnvVars = {"ANDROID_HOME", "ANDROID_SDK_ROOT"};
```

Comment:
So this introduces `$ANDROID_HOME` as a place for finding both the Android SDK & NDK, correct?

Are we worried this could break existing users? I could see users having `$ANDROID_HOME` set from Android Studio or something else, but Xamarin.Android is currently using the Android SDK installed by Visual Studio . A different SDK/NDK could get selected?

If we do this, should we put the env var checks last?

Reply:
`$ANDROID_HOME` and `$ANDROID_SDK_ROOT` are checked, yes. Without these I wasn't able to make it detect the SDK instance I wanted it to use - it searched the `$PATH`. This addition will not affect users who set their preferred NDK location from the IDE, as that is still the first path checked. Furthermore, we now collect all the NDK paths and check their versions, then return the newest compatible one - that's the main goal of this PR. Environment variables should, generally, be considered before anything else, but it might break backward compatibility - it was an oversight in the original implementation of the code.

Note also that, currently, the order in which Windows and Unix look for the NDK is different:

  1. Preferred location set by the IDE (both Windows and Unix)
  2. Environment variables, with this PR (both Windows and Unix)
  3. Other locations
    1. `$PATH` directories (Unix)
    2. Known hard-coded locations (Windows)
  4. Other locations
    1. Known hard-coded locations (Unix, only mac really)
    2. `$PATH` directories (Windows)
  5. All discovered SDK locations (Windows)

This needs to be fixed in some future PR, but for now - not sure where the environment variable check would have to be placed? I feel it needs to be before `$PATH`, since that variable is usually set account or system wide (e.g. in system preferences for the user/machine on  Windows, and in `/etc` and user's shell init scripts on Unix) and it's not very comfortable to manipulate it ad-hoc, unlike the easy way to set the supported environment variables.